### ISSUE TITLE
Debug Failure. False expression

### ISSUE THREAD

**Issue by ca0v** [Microsoft/TypeScript#30591]
Trying to do a composite build modeled after microsoft/typescript but generating AMD modules fails:

```
Error: Debug Failure. False expression.
    at mergeSymbol (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:26088:26)
    at H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:26187:47
    at Map.forEach (<anonymous>)
    at mergeSymbolTable (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:26185:20)
    at initializeTypeChecker (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:49309:21)
    at Object.createTypeChecker (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:25950:9)
    at Object.getTypeChecker (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:72860:79)
    at getReferencedFiles (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:74290:43)
    at Object.create (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:74329:41)
    at createBuilderProgramState (H:\code\ca0v\speech-lab\node_modules\typescript\lib\tsc.js:74572:37)
```

You can find a link to the project on https://stackoverflow.com/questions/55348498/problems-building-with-typescript-transpiler-tsc-using-the-build-flag, but basically if you set the module to "amd" it fails to transpiler (works fine when outFile -> outDir).

**Comment by RyanCavanaugh** [comment]
@sandersn A symbol table bug appears!

**Comment by sandersn** [comment]
Repro steps:

```
> git clone git@github.com:ca0v/speech-lab
> cd speech-lab
> npm install
> cd tsc
> tsc -b
```

**Comment by sandersn** [comment]
Looks like the program has duplicates in its list of source files.

**Comment by sandersn** [comment]
Adding asserts to program.ts:addFileToFilesByName moves this error a lot closer to its origin:

```ts
function addFileToFilesByName(file: SourceFile | undefined, path: Path, redirectedPath: Path | undefined) {
    Debug.assert(!filesByName.has(path), "duplicate path");
    if (redirectedPath) {
	Debug.assert(!filesByName.has(redirectedPath), "duplicate redirected path");
	filesByName.set(redirectedPath, file);
	filesByName.set(path, file || false);
    }
    else {
	filesByName.set(path, file);
    }
}
```

**Comment by sheetalkamat** [comment]
Had encountered this while working on  #29262 but when it got closed in favor of #30288 this issue was missed. f2862b2 handled this then but need to figure out what would be correct approach with new PR.

**Comment by O4epegb** [comment]
Also has this error with typescript@3.4, although I am building project through webpack, and its only happens when in watch mode, and only at first build. After initial build everything is ok, no errors. Without watch mode everything is fine too

```
Error: Debug Failure. False expression.
    at Object.first (/Users/.../web/node_modules/typescript/lib/typescript.js:1087:15)
    at getArgumentArityError (/Users/.../web/node_modules/typescript/lib/typescript.js:50157:32)
    at resolveCall (/Users/.../web/node_modules/typescript/lib/typescript.js:50297:41)
    at resolveCallExpression (/Users/.../web/node_modules/typescript/lib/typescript.js:50583:20)
    at resolveSignature (/Users/.../web/node_modules/typescript/lib/typescript.js:50874:28)
    at getResolvedSignature (/Users/.../web/node_modules/typescript/lib/typescript.js:50905:26)
    at checkCallExpression (/Users/.../web/node_modules/typescript/lib/typescript.js:50992:29)
    at checkExpressionWorker (/Users/.../web/node_modules/typescript/lib/typescript.js:53047:28)
    at checkExpression (/Users/.../web/node_modules/typescript/lib/typescript.js:52969:38)
    at checkBinaryLikeExpression (/Users/.../web/node_modules/typescript/lib/typescript.js:52330:29)
```

Not sure how else I can help
